Design/3093-graphql-defer.md

```diff
@@ -0,0 +1,356 @@
+
+```swift
+// Sample code for RequestChainNetworkTransport
+public func send<Operation: GraphQLOperation>(
+  operation: Operation,
+  cachePolicy: CachePolicy = .default,
+  contextIdentifier: UUID? = nil,
+  callbackQueue: DispatchQueue = .main,
+  completionHandler: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
+) -> Cancellable {
+  // request chain and request are built
+
+  if Operation.hasDeferredFragments {
+    request.addHeader(
+      name: "Accept",
+      value: "multipart/mixed;deferSpec=20220824,application/json"
+    )
+  }
+
+  // request chain kickoff
+}
+
+// Sample of new property on GraphQLOperation
+public protocol GraphQLOperation: AnyObject, Hashable {
+  // other properties not shown
+
+  static var hasDeferredFragments: Bool { get } // computed for each operation during codegen
+}
+```
+
+### Response parsing
+
+Apollo iOS already has support for parsing incremental delivery responses. That provides a great foundation to build on however there are some changes needed.
+
+#### Multipart parsing protocol
+
+The current `MultipartResponseParsingInterceptor` implementation is specific to the `subscriptionSpec` version `1.0` specification. Adopting a protocol with implementations for each of the supported specifications will enable us to support any number of incremental delivery specifications in the future.
+
+These would be registered with the `MultipartResponseParsingInterceptor` each with a unique specification string, to be used as a lookup key. When a response is received the specification string is extracted from the response `content-type` header, and the correct specification parser can be used to parse the response data.
+
+```swift
+// Sample code in MultipartResponseParsingInterceptor
+public struct MultipartResponseParsingInterceptor: ApolloInterceptor {
+  private static let multipartParsers: [String: MultipartResponseSpecificationParser.Type] = [
+    SubscriptionResponseParser.protocolSpec: SubscriptionResponseParser.self,
+    DeferResponseParser.protocolSpec: DeferResponseParser.self
+  ]
+
+  public func interceptAsync<Operation>(
+    chain: RequestChain,
+    request: HTTPRequest<Operation>,
+    response: HTTPResponse<Operation>?,
+    completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
+  ) where Operation : GraphQLOperation {
+    // response validators not shown
+
+    guard
+      let multipartBoundary = response.httpResponse.multipartBoundary,
+      let protocolSpec = response.httpResponse.multipartProtocolSpec,
+      let protocolParser = Self.multipartParsers[protocolSpec]
+    else {
+      // call request chain error handler
+
+      return
+    }
+
+    let dataHandler: ((Data) -> Void) = { data in
+      // proceed ahead on the request chain
+    }
+
+    let errorHandler: (() -> Void) = {
```

Comment:
We'll just need to make sure that we have adequate testing around this to ensure these closures don't cause retain cycles. Easy to capture `self` here and cause memory leaks.